### PR TITLE
[ Misc ] `fp8-marlin` channelwise via `compressed-tensors`

### PR DESCRIPTION
**Pull request by robertgshaw2-redhat**
SUMMARY:
* support `fp8_marlin` via `compressed-tensors`
* add support for `fp8_marlin` with `channelwise` scales
* testing should be covered by existing models running on Ampere, but also added a weight-only FP8 checkpoint

Evals on GSM8k for per-tensor and channelwise checkpoints:
```
vllm (pretrained=neuralmagic/Meta-Llama-3.1-8B-Instruct-FP8,max_model_len=4096), gen_kwargs: (None), limit: None, num_fewshot: 5, batch_size: auto
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.7786|±  |0.0114|
|     |       |strict-match    |     5|exact_match|↑  |0.7506|±  |0.0119|

vllm (pretrained=neuralmagic/Meta-Llama-3.1-8B-Instruct-FP8-dynamic,max_model_len=4096), gen_kwargs: (None), limit: None, num_fewshot: 5, batch_size: auto
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.7718|±  |0.0116|
|     |       |strict-match    |     5|exact_match|↑  |0.7536|±  |0.0119|

vllm (pretrained=nm-testing/Meta-Llama-3-8B-Instruct-W8A8-FP8-Channelwise-compressed-tensors,max_model_len=4096), gen_kwargs: (None), limit: None, num_fewshot: 5, batch_size: auto
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.7475|±  | 0.012|
|     |       |strict-match    |     5|exact_match|↑  |0.7483|±  | **0.012|**

vllm (pretrained=nm-testing/Qwen2-1.5B-Instruct-FP8W8,tensor_parallel_size=1,distributed_executor_backend=ray,trust_remote_code=true,max_model_len=4096), gen_kwargs: (None), limit: 1000.0, num_fewshot: 5, batch_size: auto
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value|   |Stderr|
|-----|------:|----------------|-----:|-----------|---|----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.585|±  |0.0156|
|     |       |strict-match    |     5|exact_match|↑  |0.578|±  |0.0156|
```

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

File: vllm/model_executor/layers/quantization/fp8.py [hunk 1]
```diff
@@ -169,19 +170,29 @@ def process_weights_after_loading(self, layer: Module) -> None:
...
+        # If checkpoint is fp8, handle that there are N scales for N
+        # shards in a fused module
         else:
-            # Dequant -> Quant with max scale.
-            max_w_scale, weight = requantize_with_max_scale(
-                weight=layer.weight,
-                weight_scale=layer.weight_scale,
-                logical_widths=layer.logical_widths,
-            )
+            # If using marlin (w8a16), kernel uses channelwise weights,
```
Comment (robertgshaw2-neuralmagic) [inline vllm/model_executor/layers/quantization/fp8.py:176]: since we extend to channelwise anyways, figured we might as well not requantize here

File: vllm/model_executor/layers/quantization/utils/marlin_utils_fp8.py [hunk 2]
```diff
@@ -74,10 +75,7 @@ def prepare_fp8_layer_for_marlin(layer: torch.nn.Module) -> None:
     layer.weight = torch.nn.Parameter(marlin_qweight, requires_grad=False)
 
     # WEIGHT SCALES
-    # Currently Marlin doesn't support per-tensor scales, so we
-    # expand it to channelwise
-    scales = layer.weight_scale.repeat(1, part_size_n).to(
-        layer.orig_dtype).to(device)
+    scales = layer.weight_scale.to(layer.orig_dtype)
```
Comment (robertgshaw2-neuralmagic) [inline vllm/model_executor/layers/quantization/utils/marlin_utils_fp8.py:78]: now, inputs to this will be channelwise

File: vllm/model_executor/layers/quantization/compressed_tensors/schemes/compressed_tensors_w8a16_fp8.py [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+    # So if we have a fused module (QKV, MLP) with per tensor scales (thus N
+    # scales being passed to the kernel), we requantize with a single scale.
+    def process_weights_after_loading(self, layer) -> None:
+        if self.strategy == QuantizationStrategy.TENSOR:
+            ws_channelwise = convert_to_channelwise(layer.weight_scale,
+                                                    layer.logical_widths)
+            layer.weight_scale = torch.nn.Parameter(ws_channelwise,
+                                                    requires_grad=False)
+
+        prepare_fp8_layer_for_marlin(layer, strategy="channel")
```
Comment (mgoin) [inline vllm/model_executor/layers/quantization/compressed_tensors/schemes/compressed_tensors_w8a16_fp8.py:38]: We should check that the strategy is TENSOR or CHANNEL at every usage, so we should have an else or assert case here

File: vllm/model_executor/layers/quantization/fp8.py (hunk 1 above)
Comment (mgoin) [reply vllm/model_executor/layers/quantization/fp8.py:176]: How does it expand the scales? 

If it is `layer.weight_scale = [A, B, C]` will it expand to `weight_scale = [A, A, ..., B, B, ..., C, C, ...]`?

File: vllm/model_executor/layers/quantization/utils/marlin_utils_fp8.py [hunk 4]
```diff
@@ -46,7 +46,8 @@ def apply_fp8_marlin_linear(
     return output.reshape(out_shape)
 
 
-def prepare_fp8_layer_for_marlin(layer: torch.nn.Module) -> None:
+def prepare_fp8_layer_for_marlin(layer: torch.nn.Module,
+                                 strategy: str = "tensor") -> None:
```
Comment (mgoin) [inline vllm/model_executor/layers/quantization/utils/marlin_utils_fp8.py:50]: nit: maybe we should move the CT QuantizationStrategy enum out of CT

File: vllm/model_executor/layers/quantization/utils/marlin_utils_fp8.py (hunk 2 above)
Comment (mgoin) [reply vllm/model_executor/layers/quantization/utils/marlin_utils_fp8.py:78]: Nice!

File: vllm/model_executor/layers/quantization/utils/marlin_utils_fp8.py (hunk 4 above)
Comment (mgoin) [reply vllm/model_executor/layers/quantization/utils/marlin_utils_fp8.py:50]: actually it looks like the strategy string is unused! ...for now ;)

**Review by mgoin — APPROVED** [review]

**Comment by RonanKMcGovern** [comment]
This is great, really nice addition, the fp8 is really fast. It's pretty
much as fast as NIM on batch size of 1 and only falls behind above around
b16.
